From 5de1f7da9308a4568aa4f7b829d99304a96b27bf Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Fri, 22 Nov 2024 15:33:47 +0100
Subject: [PATCH 1365/1695] drivers: stm32_rng: fix RNG conditional reset
 behavior
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The SEIS bit in RNG_SR is used to indicate to software that a seed error
has occurred. It is not mandatory to panic when SEIS is set because the
RNG hardware can automatically resolve the issue. The SECS bit is more
appropriate and is already implemented.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 927fa3b54d7538a00be06c706803212b4ed045b7 drivers: stm32_rng: fix RNG conditional reset behavior]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: I601dd97a8280e91e6d0afcb164b54b6a706c85a5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/420388
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/stm32_rng.c | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index b33a593fd..ab5bebfc5 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -125,9 +125,8 @@ static vaddr_t get_base(void)
  * Indeed, when SEIS is set and SECS is cleared it means RNG performed
  * the reset automatically (auto-reset).
  * 2. If SECS was set in step 1 (no auto-reset) wait for CONDRST
- * to be cleared in the RNG_CR register, then confirm that SEIS is
- * cleared in the RNG_SR register. Otherwise just clear SEIS bit in
- * the RNG_SR register.
+ * to be cleared in the RNG_CR register. Otherwise just clear SEIS bit
+ * in the RNG_SR register.
  * 3. If SECS was set in step 1 (no auto-reset) wait for SECS to be
  * cleared by RNG. The random number generation is now back to normal.
  */
@@ -165,10 +164,6 @@ static void conceal_seed_error_cond_reset(void)
 			return;
 		}
 
-		/* Check SEIS is cleared (step 2.) */
-		if (io_read32(rng_base + RNG_SR) & RNG_SR_SEIS)
-			panic();
-
 		/* Wait SECS is cleared (step 3.) */
 		if (io_read32(rng_base + RNG_SR) & RNG_SR_SECS) {
 			if (timed_out)
-- 
2.47.2

